macOS/unwindmacos.cpp

```diff
@@ -20,49 +20,116 @@ limitations under the License.
+    }
+  }
+}
+
+void UnwindDataMacOS::LookupLSDA(size_t original_address) {
+  if (last_lookup.IsLsdaMiss(original_address)) {
+    if (!lsda_map.empty() && last_lookup.encoding & UNWIND_HAS_LSDA) {
+      auto it = lsda_map.upper_bound(original_address);
+      if (it == lsda_map.begin()) {
+        last_lookup.SetLsda(-1, 0, it->first - 1);
+      } else if (it == lsda_map.end()) { // Sentinel entry
+        last_lookup.SetLsda(-1, prev(it)->first, -1);
+      } else {
+        last_lookup.SetLsda(prev(it)->second, prev(it)->first, it->first - 1);
+      }
     } else {
-      last_encoding_lookup = LastEncodingLookup(prev(it)->second, prev(it)->first, it->first - 1);
+      last_lookup.SetLsda(0, last_lookup.encoding_min_address, last_lookup.encoding_max_address);
     }
   }
+}
 
-  if (!last_encoding_lookup.IsValid()) {
+void UnwindDataMacOS::AddMetadata(size_t original_address, size_t translated_address) {
+  LookupEncoding(original_address);
+  if (!last_lookup.IsEncodingValid()) {
     return;
   }
 
-  if (metadata_list.empty() || metadata_list.back().encoding != last_encoding_lookup.encoding) {
-    metadata_list.push_back(Metadata(last_encoding_lookup.encoding,
+  LookupLSDA(original_address);
+  if (!last_lookup.IsLsdaValid()) {
+    return;
+  }
+
+  if (metadata_list.empty()
+      || metadata_list.back().encoding != last_lookup.encoding
+      || metadata_list.back().lsda != last_lookup.lsda) {
+    metadata_list.push_back(Metadata(last_lookup.encoding, last_lookup.lsda,
                                      translated_address, translated_address));
   } else {
     metadata_list.back().translated_max_address = translated_address;
   }
 }
 
+UnwindDataMacOS::UnwindDataMacOS() {
+  unwind_section_address = NULL;
+  unwind_section_size = 0;
+  unwind_section_buffer = NULL;
+  unwind_section_header = NULL;
+  last_lookup = LastLookup();
+}
+
+UnwindDataMacOS::~UnwindDataMacOS() {
+  free(unwind_section_buffer);
+  unwind_section_buffer = NULL;
+}
+
+void UnwindGeneratorMacOS::CheckUnwindBufferBounds(ModuleInfo *module, const char *array_description,
+                                                   size_t start_address, size_t size) {
+  UnwindDataMacOS *unwind_data = (UnwindDataMacOS *)module->unwind_data;
+
+  if (start_address < (size_t)unwind_data->unwind_section_buffer
+      || (size_t)unwind_data->unwind_section_buffer + unwind_data->unwind_section_size < start_address + size) {
+    FATAL("%s is located outside the Unwind Section buffer\n", array_description);
+  }
+}
+
+void UnwindGeneratorMacOS::SanityCheckUnwindHeader(ModuleInfo *module) {
+  UnwindDataMacOS *unwind_data = (UnwindDataMacOS *)module->unwind_data;
+  unwind_info_section_header *unwind_section_header = unwind_data->unwind_section_header;
+
+  if (unwind_section_header->version != UNWIND_SECTION_VERSION) {
+    FATAL("Unexpected version (%u) in Unwind Section Header",
+          unwind_data->unwind_section_header->version);
```

Comment:
You can skip the `unwind_data->` here as you already have a pointer to it in `unwind_section_header`.